test(freedrawing): test enhancement

ShaMan123 opened the pull request:
This PR is a byproduct of the next PR that will add erasing visual tests.
I added logic to freedrawing tests to prepare for testing erasing and decided (with your voice guiding me :)) to split up the PR into dedicated changes.
This PR adds another layer of tests to free drawing by testing visuals before and after mouseup and comparing between them.
I think it's relevant and to the point.
It is proven by the tests failing (freedrawing4 - `Simple free drawing, with opacity`) - ~~you might want to output the mesh result to a different file to compare changes. It might be desirable in general so tell me if to commit a different golden name for the mesh test.~~ **Shadow + opacity cause this** it seems

ShaMan123 commented:
@asturur please take a look at c6f33d2c4a6c1670107f2329e649002b1842429c
I reverted because so many tests began to fail... I am clueless as to why. I tried to refactor the code w/o changing logic to expose a function to compare to visual objects I want for freedrawing/erasing

ShaMan123 commented:
Regarding Eraser tests...
If I add them I need to export to the website a build including the Eraser which makes a lot of unit tests fail because of `Object#erasable = true` by default...
What do you suggest?

asturur commented on test/visual/freedraw.js:
```diff
@@ -27,178 +27,2225 @@
+      "x": 164.343,
+      "y": 23.064
+    }
+  ];
```
this pr seems all good, can you just keep this in a line for scrolling sanity only?
Even if we make a string and then we parse it each time.

ShaMan123 replied:
The IDE I use (VSCODE) allows me to collapse it with ease. That's one reason I changed it.
I prefer to extract it to a dedicated json file. Do you have any objections? It's better IMO. Once we migrate importing will be simple.

asturur replied:
i have VSCODE too, but yes it can be on its own file, too, as every fixture usually is.

ShaMan123 commented:
What about the failing test?

asturur commented:
Is the broken test a new test?
i see it has opacity and shadow, was the golden built with node?

asturur commented:
The website visual test got broken, probably the qunit update.
I need to fix that so i can see the difference in chrome/firefox

ShaMan123 commented:
> i see it has opacity and shadow, was the golden built with node?

The test was built in node. It is a new test. It's the upper canvas test.

asturur commented:
We can fix the the display in browser, take the golden from there, and disable the node test once we understand how much the difference is. With shadow and opacity is mostly normal

ShaMan123 commented:
I have just updated ruby so I can't serve the website as well.
Makes sense to disable the test on node as it is interactive visuals only.

ShaMan123 commented:
I have managed to serve the website

expected is golden generated in node 
actual is generated in browser (chrome)
![image](https://user-images.githubusercontent.com/34343793/169282256-02052dc0-99fb-4204-8d88-dd64e0c3ca3a.png)

ShaMan123 commented:
If I understand correctly node canvas doesn't apply opacity to the shadow...
What is the expected behavior?

asturur commented:
we can take the browser one that is most common case ( especially for free drawing ) and we can disable the test when we detect node with fabric.isLikelyNode and move forward.
Did you manage to fix the qunit display in the website?
I remember i worked on it really recently and i fixed it.

asturur commented:
<img width="475" alt="image" src="https://user-images.githubusercontent.com/1194048/169283104-10e56553-ea51-4cb2-8ee5-fa6297e78f58.png">

i fixed it here, then:

<img width="450" alt="image" src="https://user-images.githubusercontent.com/1194048/169283175-514dd903-d0dc-4f34-b393-8710201baa05.png">

i broke it there and i didn't verify it.

I saw we updated qunit in fabric and i thought it was a good idea to have the same version

asturur commented:
When hiding passed tests is still broken, anyway, there are 22 failing tests in 5.2.0 mostly on svg import, i wonder what happened

ShaMan123 commented:
I don't see any of it...
Did you run the `start` cmd from fabric.js dir?


![image](https://user-images.githubusercontent.com/34343793/169284690-d2610dac-6961-4fb8-974d-2f7b23946003.png)

ShaMan123 commented:
> <img alt="image" width="475" src="https://user-images.githubusercontent.com/1194048/169283104-10e56553-ea51-4cb2-8ee5-fa6297e78f58.png">
> 
> i fixed it here, then:
> 
> <img alt="image" width="450" src="https://user-images.githubusercontent.com/1194048/169283175-514dd903-d0dc-4f34-b393-8710201baa05.png">
> 
> i broke it there and i didn't verify it.
> 
> I saw we updated qunit in fabric and i thought it was a good idea to have the same version

Where is this? In the website? I have the website up and running locally

ShaMan123 commented:
All set

ShaMan123 commented:
I forget about the json...
I guess it will wait for es6
